File: plugin/src/ios/FirebasePlugin.m
#import "FirebasePlugin.h"
#import "Firebase.h"
#import "AppDelegate.h"
#import <Cordova/CDVPlugin.h>

@implementation FirebasePlugin

- (void)analyticsLogEvent:(CDVInvokedUrlCommand *)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* name = [command.arguments objectAtIndex:0];
    NSDictionary *parameters;

    @try {
        NSString *description = NSLocalizedString([command argumentAtIndex:1 withDefault:@"No Message Provided"], nil);

        parameters = @{ NSLocalizedDescriptionKey: description };
    }
    @catch (NSException *execption) {
        parameters = [command argumentAtIndex:1];
    }

    [self.commandDelegate runInBackground:^{
        [FIRAnalytics logEventWithName:name parameters:parameters];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)analyticsSetUserId:(CDVInvokedUrlCommand *)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* userId = [command.arguments objectAtIndex:0];

    [FIRAnalytics setUserID:userId];

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)analyticsSetScreenName:(CDVInvokedUrlCommand *)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* name = [command.arguments objectAtIndex:0];

    [FIRAnalytics setScreenName:name screenClass:NULL];

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)analyticsSetUserProperty:(CDVInvokedUrlCommand *)command {
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    NSString* name = [command.arguments objectAtIndex:0];
    NSString* value = [command.arguments objectAtIndex:1];

    [FIRAnalytics setUserPropertyString:name forName:value];

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)remoteConfigFetch:(CDVInvokedUrlCommand *)command {
    long expirationDuration = [[command argumentAtIndex:43200] longValue];

    [self.remoteConfig fetchWithExpirationDuration:expirationDuration completionHandler:^(FIRRemoteConfigFetchStatus status, NSError *error) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

        if (error) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)remoteConfigSetup:(CDVInvokedUrlCommand *)command {
    long expirationDuration = [[command argumentAtIndex:43200] longValue];

    FIRRemoteConfigSettings *remoteConfigSettings = [[FIRRemoteConfigSettings alloc] init];

    self.remoteConfig = [FIRRemoteConfig remoteConfig];
    self.remoteConfig.configSettings = remoteConfigSettings;
    self.remoteConfig.configSettings.minimumFetchInterval = expirationDuration;

    [self.remoteConfig fetchAndActivateWithCompletionHandler:^(FIRRemoteConfigFetchAndActivateStatus status, NSError * _Nullable error) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

        if (error) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)remoteConfigGetArray:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        FIRRemoteConfigValue* configValue = [self getConfigValue:command];

        if (configValue != nil) {
            NSError *error = nil;
            NSArray *jsonArray = [NSJSONSerialization JSONObjectWithData:configValue.dataValue options:kNilOptions error:&error];

            if (jsonArray == nil) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:@[]];
            } else {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:jsonArray];
            }
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)remoteConfigGetBoolean:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        FIRRemoteConfigValue* configValue = [self getConfigValue:command];

        if (configValue != nil) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:configValue.boolValue];
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)remoteConfigGetNumber:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        FIRRemoteConfigValue* configValue = [self getConfigValue:command];

        if (configValue != nil) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDouble:[configValue.numberValue doubleValue]];
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)remoteConfigGetString:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        FIRRemoteConfigValue* configValue = [self getConfigValue:command];

        if (configValue != nil) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:configValue.stringValue];
        }

        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (FIRRemoteConfigValue*)getConfigValue:(CDVInvokedUrlCommand *)command {
    NSString* key = [command argumentAtIndex:0];

    return [self.remoteConfig configValueForKey:key];
}

@end

File: plugin/src/ios/FirebasePlugin.swift
import Firebase

@objc(FirebasePlugin)
class FirebasePlugin : CDVPlugin {
    var dynamicLinkCallback = ""
    var remoteConfig: RemoteConfig!

    override func pluginInitialize() {
        DynamicLinks.performDiagnostics(completion: nil)
    }

    @objc(analyticsLogEvent:)
    private func analyticsLogEvent(command: CDVInvokedUrlCommand) {
        let name = command.arguments[0] as? String ?? ""
        let parameters = command.arguments[1] as? [String : Any]

        Analytics.logEvent(name, parameters: parameters)

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK), callbackId: command.callbackId)
    }

    @objc(analyticsSetScreenName:)
    private func analyticsSetScreenName(command: CDVInvokedUrlCommand) {
        let name = command.arguments[0] as? String ?? ""

        Analytics.logEvent(AnalyticsEventScreenView, parameters: [AnalyticsParameterScreenName: name])

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK), callbackId: command.callbackId)
    }

    @objc(analyticsSetUserId:)
    private func analyticsSetUserId(command: CDVInvokedUrlCommand) {
        let userId = command.arguments[0] as? String ?? ""

        Analytics.setUserID(userId)

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK), callbackId: command.callbackId)
    }

    @objc(analyticsSetUserProperty:)
    private func analyticsSetUserProperty(command: CDVInvokedUrlCommand) {
        let name = command.arguments[0] as? String ?? ""
        let value = command.arguments[0] as? String ?? ""

        Analytics.setUserProperty(name, forName: value)

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK), callbackId: command.callbackId)
    }

    @objc(dynamicLinksData:)
    private func dynamicLinksData(command: CDVInvokedUrlCommand) {
        dynamicLinkCallback = command.callbackId
    }

    @objc(remoteConfigFetch:)
    private func remoteConfigFetch(command: CDVInvokedUrlCommand) {
        remoteConfig.fetch() { (status, error) -> Void in
          if status == .success {
            self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK), callbackId: command.callbackId)
          } else {
            self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_ERROR), callbackId: command.callbackId)
          }
        }
    }

    @objc(remoteConfigSetup:)
    private func remoteConfigSetup(command: CDVInvokedUrlCommand) {
        let interval = command.arguments[0] as? Double ?? 43200
        let settings = RemoteConfigSettings()

        settings.minimumFetchInterval = interval

        remoteConfig = RemoteConfig.remoteConfig()
        remoteConfig.configSettings = settings
        remoteConfig.fetch() { (status, error) -> Void in
          if status == .success {
            self.remoteConfig.activate() { (changed, error) in
                self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK), callbackId: command.callbackId)
            }
          } else {
            self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_ERROR), callbackId: command.callbackId)
          }
        }
    }

    @objc(remoteConfigGetArray:)
    private func remoteConfigGetArray(command: CDVInvokedUrlCommand) {
        let name = command.arguments[0] as? String ?? ""
        var value: [AnyHashable] = []

        do {
            value = try JSONSerialization.jsonObject(with: remoteConfig.configValue(forKey: name).dataValue, options: []) as? [AnyHashable] ?? []
        } catch {
            value = []
        }

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: value), callbackId: command.callbackId)
    }

    @objc(remoteConfigGetBoolean:)
    private func remoteConfigGetBoolean(command: CDVInvokedUrlCommand) {
        let name = command.arguments[0] as? String ?? ""
        let value = remoteConfig.configValue(forKey: name).boolValue

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: value), callbackId: command.callbackId)
    }

    @objc(remoteConfigGetNumber:)
    private func remoteConfigGetNumber(command: CDVInvokedUrlCommand) {
        let name = command.arguments[0] as? String ?? ""
        let value = remoteConfig.configValue(forKey: name).numberValue as! Double

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: value), callbackId: command.callbackId)
    }

    @objc(remoteConfigGetString:)
    private func remoteConfigGetString(command: CDVInvokedUrlCommand) {
        let name = command.arguments[0] as? String ?? ""
        let value = remoteConfig.configValue(forKey: name).stringValue

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: value), callbackId: command.callbackId)
    }

    func application(_ application: UIApplication, continue userActivity: NSUserActivity,
                     restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
      let handled = DynamicLinks.dynamicLinks().handleUniversalLink(userActivity.webpageURL!) { (dynamiclink, error) in
        let url = dynamiclink?.url ?? ""

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: url), callbackId: command.callbackId)
      }

      return handled
    }

    @available(iOS 9.0, *)
    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any]) -> Bool {
      return application(app, open: url,
                         sourceApplication: options[UIApplication.OpenURLOptionsKey.sourceApplication] as? String,
                         annotation: "")
    }

    func application(_ application: UIApplication, open url: URL, sourceApplication: String?, annotation: Any) -> Bool {
      if let dynamicLink = DynamicLinks.dynamicLinks().dynamicLink(fromCustomSchemeURL: url) {
        let url = dynamicLink.url ?? ""

        self.commandDelegate.send(CDVPluginResult(status: CDVCommandStatus_OK, messageAs: url), callbackId: command.callbackId)

        return true
      }

      return false
    }
}
